projects/cm_mcu/clocksynth.c

```diff
@@ -7,155 +7,64 @@
-    return status;
-  // Make clock buffer for xcvrs pick synthesized clock
-  status = apollo_i2c_ctl_reg_w(CLOCK_I2C_BASE, CLOCK_WRITE_EXPANDER_I2C_ADDRESS, 1, CLOCK_EXPANDER_OUTPUT_PORT_0, 1,
-                                CLOCK_EXPANDER_CHOOSE_CLOCKSYNTH_4XCVR);
-  if (status != 0)
-    return status;
-  // Configuring Clock Synthesizer chip (enable and reset) via expander
-  status = apollo_i2c_ctl_reg_w(CLOCK_I2C_BASE, CLOCK_WRITE_EXPANDER_I2C_ADDRESS, 1, CLOCK_EXPANDER_OUTPUT_PORT_1, 1,
-                                CLOCK_EXPANDER_ENABLE_CLOCKSYNTH);
-  if (status != 0)
-    return status;
-  status = apollo_i2c_ctl_reg_w(CLOCK_I2C_BASE, CLOCK_WRITE_EXPANDER_I2C_ADDRESS, 1, CLOCK_EXPANDER_OUTPUT_PORT_1, 1,
-                                CLOCK_EXPANDER_RESET_CLOCKSYNTH);
-  if (status != 0)
-    return status;
-  // Clear sticky flags of clock synth status monitor (raised high after reset)
-  status = apollo_i2c_ctl_reg_w(CLOCK_I2C_BASE, CLOCK_SYNTH_I2C_ADDRESS, 1, CLOCK_SYNTH_STICKY_FLAG_REGISTER, 1, 0);
-  return status;
-}
-
-static int write_register(int RegList[][2], int n_row)
-{
-  bool ChangePage;
-  int HighByte = -1;
-  int status = -10;
-  for (int i = 0; i < n_row; i++) {
-    int NewHighByte = RegList[i][0] >> 8; // most significant 8 bits, 16 bits in total
-    if (NewHighByte != HighByte) {
-      ChangePage = true;
+  while (getPowerControlState() != POWER_ON) {
+    vTaskDelay(pdMS_TO_TICKS(10)); // delay 10 ms
+  }
+  int status = -99;
+  // Clear sticky flags of clock synth 5341 status monitor (raised high after reset)
+  for (uint8_t i = 0; i < NSUPPLIES_CLKR0A; i++) {
+    uint8_t data = 0x1U << clkr0a_moni2c_addrs[i].mux_bit;
+    int res = apollo_i2c_ctl_w(CLOCK_I2C_BASE, clkr0a_moni2c_addrs[i].mux_addr, 1, data);
+    if (res != 0) {
+      log_warn(LOG_SERVICE, "Mux write error %s, break (instance=%s,ps=%d)\r\n", SMBUS_get_error(res), clkr0a_moni2c_addrs[i].name, i);
     }
-    else {
-      ChangePage = false;
+    for (uint8_t c = 0; c < NCOMMANDS_FLG_CLKR0A; c++) {
+      uint8_t page_reg_value = clockr0a_args.commands[clockr0a_args.n_commands - c - 1].page;
+      int res = apollo_i2c_ctl_reg_w(CLOCK_I2C_BASE, CLOCK_SYNTH5341_I2C_ADDRESS, 1, CLOCK_CHANGEPAGE_REG_ADDR, 1, page_reg_value);
+      if (res != 0) {
+        log_error(LOG_SERVICE, "%s: page fail %s\r\n", clockr0a_args.name, SMBUS_get_error(res));
+      }
+      status = apollo_i2c_ctl_reg_w(CLOCK_I2C_BASE, CLOCK_SYNTH5341_I2C_ADDRESS, 1, clockr0a_args.commands[clockr0a_args.n_commands - c - 1].command, 1, 0);
+      if (status != 0)
+        return status;
     }
-    HighByte = NewHighByte;
-    uint8_t LowByte = RegList[i][0] - (NewHighByte << 8);
-    uint16_t LowByte_reg_addr = LowByte;
+  }
 
-    if (ChangePage) {
-      status = apollo_i2c_ctl_reg_w(CLOCK_I2C_BASE, CLOCK_SYNTH_I2C_ADDRESS, 1, CLOCK_CHANGEPAGE_REG_ADDR, 1, NewHighByte);
+  // Clear sticky flags of clock synth 5395 status monitor (raised high after reset)
```

Comment:
can you rewrite this code so that there is less copy-pasta? It looks like this code is identical to the loop above for the 5341.

Reply:
Done.